# Neural Machine Translation Tutorial

### myContradict Dataset: [https://github.com/ye-kyaw-thu/myContradict](https://github.com/ye-kyaw-thu/myContradict)  
### NMT Framework: [https://github.com/marian-nmt/marian](https://github.com/marian-nmt/marian)  

By Ye Kyaw Thu, Lab Leader, Language Understaning Lab, AI Mario, Myanmar

## Data Preparation

In [1]:
%pwd

'/mnt/disk1/ye/exp/contra'

In [2]:
!git clone https://github.com/ye-kyaw-thu/myContradict

Cloning into 'myContradict'...
remote: Enumerating objects: 158, done.
remote: Counting objects: 100% (158/158), done.
remote: Compressing objects: 100% (149/149), done.
remote: Total 158 (delta 45), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (158/158), 2.27 MiB | 3.09 MiB/s, done.
Resolving deltas: 100% (45/45), done.


In [3]:
%cd myContradict/

/mnt/disk1/ye/exp/contra/myContradict


In [4]:
!ls

README.md  version1.0


In [5]:
%cd version1.0

/mnt/disk1/ye/exp/contra/myContradict/version1.0


In [6]:
!tree

.
├── baseline_experiments
│   ├── openNMT_baseline.ipynb
│   └── yaml_files
│       ├── yaml_original
│       │   ├── syl_s2s.yaml
│       │   ├── syl.transformer.yaml
│       │   ├── word_s2s.yaml
│       │   └── word.transformer.yaml
│       └── yaml_pos
│           ├── syl_s2s.yaml
│           ├── syl.transformer.yaml
│           ├── word_s2s.yaml
│           └── word.transformer.yaml
├── myContradict_v1_word.txt
├── syllable_segmented
│   ├── syl_test.src.txt
│   ├── syl_test.tgt.txt
│   ├── syl_train.src.txt
│   ├── syl_train.tgt.txt
│   ├── syl_valid.src.txt
│   └── syl_valid.tgt.txt
├── with_POS_Tags
│   ├── syllable_level_segmentation
│   │   ├── syl_test.src.txt.TAGGED.TAGGED
│   │   ├── syl_train.src.txt.TAGGED.TAGGED
│   │   └── syl_valid.src.txt.TAGGED.TAGGED
│   └── word_level_segmentation
│       ├── word_test.src.txt.TAGGED.TAGGED
│       ├── word_train.src.txt.TAGGED.TAGGED
│       └── word_valid.src.txt.TAGGED.TAGGED
└── word_segmented
    ├── word_test.src.txt
    ├─

## Syllable Segmented Data

In [7]:
!head ./syllable_segmented/syl_train.src.txt

ဆွယ် တာ အ သီး မ ထ ပါ ။
ခံ တွင်း မ ပေါက် ဘူး ။
အဲ ဒီ မြင် ကွင်း က ဘယ် လောက် ရယ် စ ရာ မ ကောင်း လဲ ဆို ရင် ရယ် ချင် စိတ် ကို ချိုး နှိမ် ထား လို့ ရ တယ် ။
ခင် ဗျား နောက် ရက် အ နည်း ငယ် ကြာ မှ သွား လို့ ရ မယ် ။
ခြံ စည်း ရိုး အောက် က နေ ငုံ့ ဝင် သွား ကြ တယ် ။
ကို ငြိမ်း ရေ လူ တွေ သိပ် ချောင် နေ လို့ ထွက် မ ရ ဘဲ နေ ဦး မယ် ။
အ ခု မှ လာ ပြီး နောင် တ ရ လည်း အ ပို ပဲ ။
ဆက် တိုက် လာ သည့် များ ကို ပို ၍ သေး ငယ် သည့် ယူ နစ် များ အ ဖြစ် ပေါင်း စပ် ပေး ခြင်း ဖြင့် ပို ၍ အ ဓိပ္ပါယ် ပြည့် ဝ သော သ တင်း အ ချက် အ လက် များ ကို ရ ရှိ သည် ဥ ပ မာ အား ဖြင့် တစ် ခု ကို ဆက် တိုက် လာ သည့် ရှစ် ခု ဖြင့် ဖွဲ့ စည်း ထား သည် ။
မြေ ပြင် ဝန် ထမ်း များ သည် အ စား အ စာ ၊ ရေ နှင့် အ ခြား အ စား အ စာ မ ဟုတ် သည့် အ ရာ များ ကို သိမ်း ဆည်း ရန် လုပ် နေ သည် ။
သူ က လွယ် လွယ် ကူ ကူ ရ ထား တဲ့ အ လုပ် ကို ဆက် လုပ် ဖို့ သိပ် အ ဆင် မ ပြေ ဘူး ။


In [8]:
!head ./syllable_segmented/syl_train.tgt.txt

ဆွယ် တာ က အ သီး ထ နေ ပါ ပေါ့ ။
ခံ တွင်း ပေါက် သည် ။
အဲ ဒီ မြင် ကွင်း က ဘယ် လောက် ရယ် စ ရာ ကောင်း လဲ ဆို ရင် ရယ် ချင် စိတ် ကို ချိုး နှိမ် ထား လို့ မ ရ ဘူး ။
ခင် ဗျား အ ခု ချက် ချင်း သွား လို့ ရ တယ် ။
ခြံ စည်း ရိုး ပေါ် က နေ ကျော် တက် သွား ကြ တယ် ။
ကို ငြိမ်း ရေ လူ တွေ သိပ် ကြပ် နေ လို့ ထွက် မ ရ ဘဲ နေ ဦး မယ် ။
အ ခု မှ လာ ပြီး နောင် တ ရ လည်း အ ဆင် ပြေ ပါ သေး တယ် ။
ဆက် တိုက် လာ သည့် များ ကို ပို ၍ ကြီး မား သည့် ယူ နစ် များ အ ဖြစ် ပေါင်း စပ် ပေး ခြင်း ဖြင့် ပို ၍ အ ဓိပ္ပါယ် ပြည့် ဝ သော သ တင်း အ ချက် အ လက် များ ကို ရ ရှိ သည် ဥ ပ မာ အား ဖြင့် တစ် ခု ကို ဆက် တိုက် လာ သည့် ရှစ် ခု ဖြင့် ဖွဲ့ စည်း ထား သည် ။
မြေ ပြင် ဝန် ထမ်း များ သည် အ စား အ စာ ၊ ရေ နှင့် အ ခြား အ စား အ စာ မ ဟုတ် သည့် အ ရာ များ ကို ဖြန့် ဖြူး ရန် လုပ် နေ သည် ။
သူ က ခဲ ရာ ခဲ ဆစ် ရ ထား တဲ့ အ လုပ် ကို ဆက် လုပ် ဖို့ သိပ် အ ဆင် မ ပြေ ခဲ့ ဘူး ။


In [10]:
!wc ./syllable_segmented/*src.txt

   1214   18537  175215 ./syllable_segmented/syl_test.src.txt
   9702  144078 1357932 ./syllable_segmented/syl_train.src.txt
   1214   17782  168043 ./syllable_segmented/syl_valid.src.txt
  12130  180397 1701190 total


In [11]:
!wc ./syllable_segmented/*tgt.txt

   1214   18515  174875 ./syllable_segmented/syl_test.tgt.txt
   9702  144108 1358601 ./syllable_segmented/syl_train.tgt.txt
   1214   17778  167754 ./syllable_segmented/syl_valid.tgt.txt
  12130  180401 1701230 total


## Word Segmented Data

In [12]:
!head ./word_segmented/word_train.src.txt

ဆွယ်တာ အသီး မ ထ ပါ ။
ခံတွင်း မ ပေါက် ဘူး ။
အဲဒီ မြင်ကွင်း က ဘယ်လောက် ရယ် စရာ မ ကောင်း လဲ ဆို ရင် ရယ် ချင် စိတ် ကို ချိုးနှိမ် ထား လို့ ရ တယ် ။
ခင်ဗျား နောက် ရက် အနည်းငယ် ကြာ မှ သွား လို့ ရ မယ် ။
ခြံစည်းရိုး အောက် က နေ ငုံ့ဝင် သွားကြ တယ် ။
ကိုငြိမ်း ရေ လူ တွေ သိပ် ချောင် နေ လို့ ထွက် မ ရ ဘဲ နေ ဦး မယ် ။
အခု မှ လာ ပြီး နောင်တရ လည်း အပို ပဲ ။
ဆက်တိုက် လာ သည့် များ ကို ပို ၍ သေးငယ် သည့် ယူနစ် များ အဖြစ် ပေါင်းစပ် ပေး ခြင်း ဖြင့် ပို ၍ အဓိပ္ပါယ် ပြည့်ဝ သော သတင်း အချက်အလက် များ ကို ရရှိ သည် ဥပမာ အားဖြင့် တစ် ခု ကို ဆက်တိုက် လာ သည့် ရှစ် ခု ဖြင့် ဖွဲ့စည်း ထား သည် ။
မြေပြင် ဝန်ထမ်း များ သည် အစားအစာ ၊ ရေ နှင့် အခြား အစားအစာ မ ဟုတ် သည့် အရာ များ ကို သိမ်းဆည်း ရန် လုပ် နေ သည် ။
သူ က လွယ်လွယ်ကူကူ ရ ထား တဲ့ အလုပ် ကို ဆက်လုပ် ဖို့ သိပ် အဆင်မပြေ ဘူး ။


In [13]:
!head ./word_segmented/word_train.tgt.txt

ဆွယ်တာ က အသီး ထ နေ ပါ ပေါ့ ။
ခံတွင်းပေါက် သည် ။
အဲဒီ မြင်ကွင်း က ဘယ်လောက် ရယ် စရာ ကောင်း လဲ ဆို ရင် ရယ် ချင် စိတ် ကို ချိုးနှိမ် ထား လို့ မ ရ ဘူး ။
ခင်ဗျား အခု ချက်ချင်း သွား လို့ ရ တယ် ။
ခြံစည်းရိုး ပေါ် က နေ ကျော်တက် သွားကြ တယ် ။
ကိုငြိမ်း ရေ လူ တွေ သိပ် ကြပ် နေ လို့ ထွက် မ ရ ဘဲ နေ ဦး မယ် ။
အခု မှ လာ ပြီး နောင်တရ လည်း အဆင်ပြေ ပါ သေး တယ် ။
ဆက်တိုက် လာ သည့် များ ကို ပို ၍ ကြီးမား သည့် ယူနစ် များ အဖြစ် ပေါင်းစပ် ပေး ခြင်း ဖြင့် ပို ၍ အဓိပ္ပါယ် ပြည့်ဝ သော သတင်း အချက်အလက် များ ကို ရရှိ သည် ဥပမာ အားဖြင့် တစ် ခု ကို ဆက်တိုက် လာ သည့် ရှစ် ခု ဖြင့် ဖွဲ့စည်း ထား သည် ။
မြေပြင် ဝန်ထမ်း များ သည် အစားအစာ ၊ ရေ နှင့် အခြား အစားအစာ မ ဟုတ် သည့် အရာ များ ကို ဖြန့်ဖြူး ရန် လုပ် နေ သည် ။
သူ က ခဲရာခဲဆစ် ရ ထား တဲ့ အလုပ် ကို ဆက်လုပ် ဖို့ သိပ် အဆင်မပြေ ခဲ့ ဘူး ။


In [14]:
!wc ./word_segmented/*.src.txt

   1214   13524  170232 ./word_segmented/word_test.src.txt
   9702  105208 1319543 ./word_segmented/word_train.src.txt
   1214   13139  163441 ./word_segmented/word_valid.src.txt
  12130  131871 1653216 total


In [15]:
!wc ./word_segmented/*.tgt.txt

   1214   13529  169938 ./word_segmented/word_test.tgt.txt
   9702  105198 1320154 ./word_segmented/word_train.tgt.txt
   1214   13121  163137 ./word_segmented/word_valid.tgt.txt
  12130  131848 1653229 total


## Marian NMT Framework

အဲဒီတုန်းက OpenNMT ကို သုံးခဲ့ပေမဲ့ ဒီတခါတော့ Marian NMT Framework ကို သုံးပြီး experiment တချို့ လုပ်ကြရအောင်။  

ဆရာ့ စက်ထဲမှာက installation လုပ်ထားပြီးသားမို့ တန်း run လို့ ရလိမ့်မယ်။
မင်းတို့ စက်ထဲမှာ မရှိရင် marian setup လုပ်ကြရလိမ့်မယ်။  
Marian က C++ နဲ့ ရေးထားတာမို့ training/testing လုပ်ရင် အများကြီး မြန်ပါတယ်။  

In [16]:
!marian --help

Marian: Fast Neural Machine Translation in C++
Usage: marian [OPTIONS]

General options:
  -h,--help                             Print this help message and exit
  --version                             Print the version number and exit
  --authors                             Print list of authors and exit
  --cite                                Print citation and exit
  --build-info TEXT                     Print CMake build options and exit. Set to 'all' to print advanced options
  -c,--config VECTOR ...                Configuration file(s). If multiple, later overrides earlier
  -w,--workspace INT=2048               Preallocate arg MB of work space. Negative `--workspace -N` value allocates workspace as total available GPU memory minus N megabytes.
  --log TEXT                            Log training process information to file given by arg
  --log-level TEXT=info                 Set verbosity level of logging: trace, debug, info, warn, err(or), critical, off
  --log-time-zone TEXT  

## Quick Start


ပထမဆုံး အသုံးပြုမယ့် သူတွေအတွက်က အောက်ပါ quick start, example  လင့်တွေက အသုံးဝင်ပါလိမ့်မယ်။ Marian က GPU မရှိလည်း training/testing လုပ်လို့ ရပါတယ်။ သို့သော် CPU ပေါ်မှာ ဆိုရင်တော့ training time က ကြာနိုင်တယ်။  

Quick start link: [https://marian-nmt.github.io/quickstart/](https://marian-nmt.github.io/quickstart/)  
Examples: [https://marian-nmt.github.io/examples/](https://marian-nmt.github.io/examples/)  

## Vocab Building


Neural Network ကို သုံးမှာမို့ ထုံးစံအတိုင်း vocab ဖိုင်တွေကို ပြင်ဆင်ကြရပါတယ်။  
Git clone လုပ်ယူထားတဲ့ myContradict ဒေတာထဲက syllable unit, word unit ဖြတ်ထားတဲ့ နှစ်မျိုးစလုံးအတွက် vocab ဆောက်ကြရအောင်။ ဘာကြောင့်လဲ ဆိုတော့ ဒီနေ့ syllable နဲ့ရော word နဲ့ရော အနည်းဆုံး NMT model ဆောက်ပြီး ရလဒ်တွေကို နှိုင်းယှဉ်ကြည့်ချင်လို့။  

In [18]:
!marian-vocab --help

Create a vocabulary from text corpora given on STDIN
Usage: marian-vocab [OPTIONS]

Allowed options:
  -h,--help                             Print this help message and exit
  --version                             Print the version number and exit
  -m,--max-size UINT=0                  Generate only UINT most common vocabulary items

Examples:
  ./marian-vocab < text.src > vocab.yml
  cat text.src text.trg | ./marian-vocab > vocab.yml


## For Syllable Segmented Data

In [20]:
%cd /mnt/disk1/ye/exp/contra/

/mnt/disk1/ye/exp/contra


In [21]:
!mkdir data

In [23]:
%cd /mnt/disk1/ye/exp/contra/myContradict/version1.0/syllable_segmented

/mnt/disk1/ye/exp/contra/myContradict/version1.0/syllable_segmented


In [24]:
!ls

syl_test.src.txt  syl_train.src.txt  syl_valid.src.txt
syl_test.tgt.txt  syl_train.tgt.txt  syl_valid.tgt.txt


In [25]:
!time cat syl_test.src.txt  syl_test.tgt.txt  syl_train.src.txt  syl_train.tgt.txt  syl_valid.src.txt  syl_valid.tgt.txt | marian-vocab > vocab.syl.yml

[2026-09-26 07:29:33] Creating vocabulary...
[2026-09-26 07:29:33] [data] Creating vocabulary stdout from stdin
[2026-09-26 07:29:33] Finished

real	0m0.295s
user	0m0.257s
sys	0m0.017s


In [26]:
!head -n 20 ./vocab.syl.yml

</s>: 0
<unk>: 1
။: 2
အ: 3
မ: 4
ရ: 5
ပါ: 6
တယ်: 7
က: 8
ကို: 9
လို့: 10
သည်: 11
ကျွန်: 12
နေ: 13
မှာ: 14
တာ: 15
ဘူး: 16
တော်: 17
များ: 18
တွေ: 19


In [27]:
!tail -n 20 ./vocab.syl.yml

မြှူ: 1673
မွ: 1674
မွေ့: 1675
မွှန်း: 1676
ယိမ်း: 1677
ယိုင်: 1678
ယီ: 1679
ယျာ: 1680
ရို့: 1681
ရှဲ: 1682
လိန်း: 1683
လျောက်: 1684
လျှို: 1685
လျှို့: 1686
လျှံ: 1687
လှန့်: 1688
လှိမ့်: 1689
လှိုက်: 1690
သိမ့်: 1691
အိ: 1692

In [37]:
!wc ./vocab.syl.yml

 1692  3386 30552 ./vocab.syl.yml


In [38]:
!mv vocab.syl.yml /mnt/disk1/ye/exp/contra/data

## For Word Segmented Data

ဒီတခါတော့ word unit အတွက် vocab ဆောက်မယ်။  

In [29]:
%cd /mnt/disk1/ye/exp/contra/myContradict/version1.0/word_segmented/

/mnt/disk1/ye/exp/contra/myContradict/version1.0/word_segmented


In [30]:
!ls

word_test.src.txt  word_train.src.txt  word_valid.src.txt
word_test.tgt.txt  word_train.tgt.txt  word_valid.tgt.txt


In [31]:
!time cat word_test.src.txt  word_test.tgt.txt  word_train.src.txt  word_train.tgt.txt  word_valid.src.txt  word_valid.tgt.txt |  marian-vocab > vocab.word.yml

[2026-09-26 07:32:29] Creating vocabulary...
[2026-09-26 07:32:29] [data] Creating vocabulary stdout from stdin
[2026-09-26 07:32:29] Finished

real	0m0.224s
user	0m0.220s
sys	0m0.011s


In [33]:
!head -n 20 ./vocab.word.yml

</s>: 0
<unk>: 1
။: 2
ရ: 3
တယ်: 4
ပါ: 5
မ: 6
ကို: 7
က: 8
လို့: 9
သည်: 10
မှာ: 11
ဘူး: 12
တာ: 13
နေ: 14
တွေ: 15
ကျွန်တော်: 16
တဲ့: 17
များ: 18
တို့: 19


In [34]:
!tail -n 20 ./vocab.word.yml

ဣန္ဒြေပျက်: 6834
ဥပေက္ခာပြုထား: 6835
ဦးစားမပေး: 6836
ဦးဆောင်ပြီး: 6837
ဦးမော့: 6838
ဧကန်: 6839
ဧည့်: 6840
ေတွ: 6841
ေသာ: 6842
။ဟုတ်ကဲ့: 6843
ကြပါ။: 6844
ဖြစ်သည်။: 6845
မကြိုးစား: 6846
မည်ဟု: 6847
မလိုပါ။: 6848
မလိုဘူး: 6849
မဟုတ်ပါ: 6850
မဟုတ်ပါ။: 6851
အသီးမထပါ။: 6852
၏။: 6853


အထက်မှာ မြင်ရတဲ့အတိုင်းပဲ ပုဒ်ထီး၊ ပုဒ်မ တွေက စာလုံးတွေနဲ့ ပူးနေတာတွေကို တွေ့ရပါလိမ့်မယ်။ Syllable segmentation မှာတော့ sylbreak tool or sylbreak4all tool နဲ့ ဖြတ်လိုက်ရင် ပြဿနာ မရှိပေမဲ့၊ word unit ဖြတ်ထားတဲ့ ဖိုင်မှာတော့ ဘယ်လို လုပ်ကြမလဲ စဉ်းစားကြဖို့ လိုအပ်ပါတယ်။ မဟုတ်ရင် ရလဒ်ကို လာထိခိုက်မှာမို့လို့။  

နောက်အရေးကြီးတဲ့ တချက်က myContradict ဒေတာမှာက စာကြောင်း တစ်ကြောင်းထက် ပိုတဲ့ စာကြောင်းတွေလည်း ပါဝင်ပါသေးတယ်။ ဥပမာ ။ ။ 

In [41]:
!grep -n "။ " ./word_train.src.txt | head -n 20

26:ကောင်း ပါ တယ် ။ နာရီ ရ ပါ တယ် ။
128:ဟုတ် ပါ တယ် ။ ဟို ဘက် ကို ကြွ ပါ ။
142:ဒီ နေ့ ကျွန်တော် ပျော် တယ် ခင်ဗျား ရဲ့ ဝန်ထမ်း က ကျွန်တော့် ရဲ့ ပျောက် သွား တဲ့ ကင်မရာ ကို ပြန်ပို့ ပေး တယ် ။ လုံးဝ မေ့ နိုင် မှာ မ ဟုတ် ဘူး ။
179:ကျွန်တော် တော့ လေ့လာ လို့ အားမရ သေး ဘူး ။ နောက် တစ် ခေါက် လာ ဦး မယ် ။
201:အေး အေး ကောင်း တယ် ။ ဘယ် အရွယ် က တည်း က စာမဖတ် တတ် တာ လဲ ။
213:တစ်သိန်း ခွဲ ဆို ရင် ဒီ ငွေပို့လွှာ ပုံစံ ကို ဖြည့် ပါ ။ ငွေ အရေအတွက် ကို ဂဏန်း လို ပဲ ပီသ အောင် ရေး ပြီး ၊ စာ လို ရေး ရန် မ လို ပါ ။
277:အံ့ဩ စရာ မ ဟုတ် ပါ ဘူး ၊ ရွှီးယွင် ။ ခင်ဗျား ဘာ ပြော နေ တာ လဲ ။
305:သူ တို့ က တော့ ပြန် မ စ ကြ ဘူး ။ မြင် တယ် မ ဟုတ် လား ။
403:တဆိတ်လောက် ။ ဒီ စာအုပ် တွေ ဘယ်လောက် ကြာကြာ ယူ ထား လို့ ရ မလဲ ။
406:ရ ပါ တယ် ရှင် ။ ဒါ က မိတ်ဆွေ လို ချင် တဲ့ စာအုပ် ။
419:ရ ပါ တယ် ။ မန္တလေး တောင် ပါ တစ်ခါတည်း တက် မယ် လေ ။
425:သူ တို့ က လည်း ပြန် စ ကြ တယ် ။ မြင် တယ် မ ဟုတ် လား ။
480:ဟုတ် တယ် နော် ။ လေ ကောင်း လေ သန့် လည်း ရ တယ် ။ အခင်းအကျင်း က လည်း သပ်ရပ် တယ် ။ ဟင်း တွေ က လည်း အရသာရှိ မဲ့ ပုံ ပဲ ။
596:အဆင်သင့် ဖြစ် ပြီ ။

## Data Cleaning

In [42]:
%cd /mnt/disk1/ye/exp/contra

/mnt/disk1/ye/exp/contra


In [43]:
!mkdir scripts


အထူးသဖြင့် punctuation mark တွေကို ဖြုတ်ဖို့၊ စာကြောင်း ရှေ့ဆုံးမှာ၊ စာကြောင်း ရဲ့ နောက်ဆုံးမှာ ရှိနေတဲ့ space တွေကို ရှင်းဖို့နဲ့ တစ်ခုထက်မက ကပ်ရိုက်ထားတဲ့ space တွေကို ရှင်းဖို့ ပိုင်သွန်ကုဒ်ကို ပြင်ဆင်ခဲ့တယ်။

In [44]:
%cd scripts

/mnt/disk1/ye/exp/contra/scripts


In [45]:
!ls ./clean-punctuation.py

./clean-punctuation.py


```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
clean-punctuation.py
Cleaning punctuation characters for Burmese or Myanmar language.
Python port of clean-punctuation.pl (Ye Kyaw Thu, 22 May 2022)

Usage:
    python3 clean-punctuation.py <input-file> > <output-file>
    cat input.txt | python3 clean-punctuation.py > output.txt
"""

import sys
import re

# Punctuation characters to remove (same set as the Perl script)
PUNCT_RE = re.compile(r'[၊။\-()\[\]/\\!?,"<>]')
MULTI_SPACE_RE = re.compile(r' +')


def clean_line(line: str) -> str:
    """Remove target punctuation, trim, and collapse multiple spaces."""
    line = PUNCT_RE.sub('', line)
    line = line.strip()
    line = MULTI_SPACE_RE.sub(' ', line)
    return line


def main() -> None:
    # Read from file argument, or fall back to stdin
    if len(sys.argv) >= 2:
        fin = open(sys.argv[1], 'r', encoding='utf-8')
        close_fin = True
    else:
        fin = sys.stdin
        close_fin = False

    try:
        for raw in fin:
            line = raw.rstrip('\n').rstrip('\r')
            if not line.strip():          # skip empty / whitespace-only lines
                continue
            cleaned = clean_line(line)
            if cleaned:                   # skip lines that became empty
                print(cleaned)
    finally:
        if close_fin:
            fin.close()


if __name__ == '__main__':
    main()
```

In [47]:
%cd /mnt/disk1/ye/exp/contra/data

/mnt/disk1/ye/exp/contra/data


In [48]:
!rm ./vocab.syl.yml

In [49]:
!mkdir preprocessing

In [85]:
%cd /mnt/disk1/ye/exp/contra/myContradict/version1.0/syllable_segmented

/mnt/disk1/ye/exp/contra/myContradict/version1.0/syllable_segmented


In [86]:
!cat syl_test.src.txt  syl_test.tgt.txt  syl_train.src.txt  syl_train.tgt.txt  syl_valid.src.txt  syl_valid.tgt.txt > /mnt/disk1/ye/exp/contra/data/preprocessing/syl_all.txt

In [87]:
%cd /mnt/disk1/ye/exp/contra/myContradict/version1.0/word_segmented/

/mnt/disk1/ye/exp/contra/myContradict/version1.0/word_segmented


In [88]:
!cat word_test.src.txt  word_test.tgt.txt  word_train.src.txt  word_train.tgt.txt  word_valid.src.txt  word_valid.tgt.txt  > /mnt/disk1/ye/exp/contra/data/preprocessing/word_all.txt

အထက်မှာ လုပ်ခဲ့တဲ့ အလုပ်က soruce, target အကုန်ပေါင်းလိုက်ပြီး ဖိုင်တစ်ဖိုင်အနေနဲ့ ဆောက်လိုက်တာပါ။ 
အဲဒါကို syllable အတွက်ကော word အတွက်ကော လုပ်ခဲ့ပြီး preprocessing/ ဖိုလ်ဒါအောက်မှာ သိမ်းထားလိုက်တယ်။  


In [89]:
%cd /mnt/disk1/ye/exp/contra/data/preprocessing

/mnt/disk1/ye/exp/contra/data/preprocessing


In [90]:
!wc *

   24260   360798  3402420 syl_all.txt
   24260   360798  3402420 syl_all.txt.clean
   24260   263719  3306445 word_all.txt
   24260   263719  3306445 word_all.txt.clean
   97040  1249034 13417730 total



အားလုံးပေါင်းထားတဲ့ ဖိုင်နှစ်ဖိုင်ကို cleaning လုပ်ပြီးမှ vocab ဆောက်မယ်။  

In [91]:
!cat syl_all.txt | python ../../scripts/clean-punctuation.py > syl_all.txt.clean

In [92]:
!cat word_all.txt | python ../../scripts/clean-punctuation.py > word_all.txt.clean

In [93]:
!wc syl_all.txt.clean

  24260  334020 3295308 syl_all.txt.clean


In [94]:
!wc word_all.txt.clean

  24260  236995 3199267 word_all.txt.clean


In [95]:
!tail -n 30 ./syl_all.txt.clean

ကား အ တွက် အာ မ ခံ လုပ် လို့ ရ လား
ယဇ် ကောင် များ ဖြင့် နတ် များ ကို ပူ ဇော် ပ သ ရ သည်
ကိုယ့် ကိုယ် ကိုယ် နှစ် သက် စိတ် တွေ ဝင် လာ လေ သည်
ကမ်း နား မှာ ထိုင် ပြီး တွေး နေ ခဲ့ တယ်
အ ကူ အ ညီ မ ရ လည်း ကျေး ဇူး တင် တာ သ ဘာ ဝ ပဲ
ခင် ဗျား ဒီ စာ ကို ဖက်စ် နဲ့ ပို့ ပေး လို့ ရ မ လား
ကမ္ဘာ့ အ ကြီး ဆုံး ပို့ ကုန် တင် ပို့ သူ နှင့် ဒု တိ ယ အ ကြီး ဆုံး သွင်း ကုန် တင် သွင်း သူ ဖြစ် လာ ခဲ့ သည်
သူ ငယ် ချင်း နဲ့ အ ချိန် ညှိ လိုက် တာ နောက် လ ကျ ရင် တောင် အ တူ တူ ခ ရီး သွား လို့ မ ရ လောက် ဘူး ထင် တယ်
ကျွန် တော် စိတ် ပြေ မိ တယ်
ကျွန် မ ဘေး မှာ တော့ သူ ခ ဏ လေး ထိုင် သွား တယ်
ကြို တင် ဖြန့် ဝေ ထား တဲ့ ဖိုင် တွေ မှာ ပါ ဝင် တဲ့ အ တိုင်း မ ဆောင် ရွက် ရ ပါ
ဘာ သာ ရေး ပွဲ တော် က ဘယ် အ ချိန် မှာ ပြီး မှာ လဲ
အ ကျင့် စာ ရိတ္တ မ ကောင်း လည်း ရ ပါ တယ်
ကိုယ် က မှန် နေ ပေ မဲ့ လည်း သူ ပြော သ မျှ တောင့် ခံ နေ ရ တယ်
ဘယ် ကိစ္စ မ ဆို နှစ် ဖက် လုံး ကြည့် ပြော လို့ မ ရ ဘူး
အိပ် စင် မ ရ နိုင် တော့ ပါ
ကျွန် တော် သိ တဲ့ ဖြစ် ပုံ ပျက် ပုံ က တော့ ဒါ ပါ ပဲ
လမ်း ခွဲ လိုက် တဲ့ ခဈ် သူ ကောင် မ လေး ကို သ တိ ရ နေ လို့ စာ အုပ် ဖတ် လည်း ခေါင်

In [96]:
!tail -n 30 ./word_all.txt.clean

ကား အတွက် အာမခံ လုပ် လို့ ရ လား
ယဇ်ကောင် များ ဖြင့် နတ် များ ကို ပူဇော်ပသ ရ သည်
ကိုယ့်ကိုယ်ကိုယ် နှစ်သက် စိတ် တွေ ဝင် လာ လေ သည်
ကမ်းနား မှာ ထိုင် ပြီး တွေး နေ ခဲ့ တယ်
အကူအညီ မ ရ လည်း ကျေးဇူးတင် တာ သဘာဝ ပဲ
ခင်ဗျား ဒီ စာ ကို ဖက်စ် နဲ့ ပို့ပေး လို့ ရ မလား
ကမ္ဘာ့ အကြီးဆုံး ပို့ကုန် တင်ပို့ သူ နှင့် ဒုတိယ အကြီးဆုံး သွင်းကုန် တင်သွင်း သူ ဖြစ် လာ ခဲ့ သည်
သူငယ်ချင်း နဲ့ အချိန်ညှိ လိုက် တာ နောက် လ ကျ ရင် တောင် အတူတူ ခရီးသွား လို့ မ ရ လောက် ဘူး ထင် တယ်
ကျွန်တော် စိတ်ပြေ မိ တယ်
ကျွန်မ ဘေးမှာ တော့ သူ ခဏ လေး ထိုင် သွား တယ်
ကြိုတင် ဖြန့်ဝေ ထား တဲ့ ဖိုင် တွေ မှာ ပါဝင် တဲ့ အတိုင်း မ ဆောင်ရွက် ရ ပါ
ဘာသာရေး ပွဲတော် က ဘယ် အချိန် မှာ ပြီး မှာ လဲ
အကျင့်စာရိတ္တ မ ကောင်း လည်း ရ ပါ တယ်
ကိုယ် က မှန် နေ ပေမဲ့ လည်း သူ ပြော သမျှ တောင့်ခံ နေ ရ တယ်
ဘယ် ကိစ္စ မဆို နှစ် ဖက် လုံး ကြည့် ပြော လို့ မ ရ ဘူး
အိပ်စင် မ ရ နိုင် တော့ ပါ
ကျွန်တော် သိ တဲ့ ဖြစ်ပုံပျက်ပုံ က တော့ ဒါ ပါ ပဲ
လမ်းခွဲ လိုက် တဲ့ ချစ်သူ ကောင်မလေး ကို သတိရ နေ လို့ စာအုပ် ဖတ် လည်း ခေါင်း ထဲ မ ရောက် ဘူး
ပို့ ရန် ကောင်ကလေး လိုအပ် ပါ တယ်
ဆိုး တာ ကွာ
မင်း ရဲ့ 

## Vocab Building with Cleaned Data

တကယ်ကတော့ syllable normalization လုပ်နိုင်ရင် ပိုကောင်းတယ်။ ဒီနေရာမှာတော့ လက်ရှိ clean လုပ်ထားတဲ့ ဒေတာနဲ့ပဲ သွားကြရအောင်။  

In [64]:
!mkdir ../vocab

In [97]:
!cat ./syl_all.txt.clean | marian-vocab >  ../vocab/syl_vocab.yml

[2026-09-26 08:54:36] Creating vocabulary...
[2026-09-26 08:54:36] [data] Creating vocabulary stdout from stdin
[2026-09-26 08:54:36] Finished


In [98]:
!cat ./word_all.txt.clean | marian-vocab >  ../vocab/word_vocab.yml

[2026-09-26 08:54:37] Creating vocabulary...
[2026-09-26 08:54:37] [data] Creating vocabulary stdout from stdin
[2026-09-26 08:54:37] Finished



ဒီတခါတော့ သုံးလို့ရမယ့် vocab ဖိုင်ကို ဆောက်ပြီးသွားပြီလို့ ယူဆပါတယ်။ တချက်တော့ head/tail command သုံးပြီး confirm လုပ်ကြရအောင်။  

In [99]:
!head -n 20 ../vocab/syl_vocab.yml

</s>: 0
<unk>: 1
အ: 2
မ: 3
ရ: 4
ပါ: 5
တယ်: 6
က: 7
ကို: 8
လို့: 9
သည်: 10
ကျွန်: 11
နေ: 12
မှာ: 13
တာ: 14
ဘူး: 15
တော်: 16
များ: 17
တွေ: 18
စ: 19


In [100]:
!tail -n 20 ../vocab/syl_vocab.yml

မြှူ: 1671
မွ: 1672
မွေ့: 1673
မွှန်း: 1674
ယိမ်း: 1675
ယိုင်: 1676
ယီ: 1677
ယျာ: 1678
ရို့: 1679
ရှဲ: 1680
လိန်း: 1681
လျောက်: 1682
လျှို: 1683
လျှို့: 1684
လျှံ: 1685
လှန့်: 1686
လှိမ့်: 1687
လှိုက်: 1688
သိမ့်: 1689
အိ: 1690

In [101]:
!head -n 20 ../vocab/word_vocab.yml

</s>: 0
<unk>: 1
ရ: 2
တယ်: 3
ပါ: 4
မ: 5
ကို: 6
က: 7
လို့: 8
သည်: 9
မှာ: 10
ဘူး: 11
တာ: 12
နေ: 13
တွေ: 14
ကျွန်တော်: 15
တဲ့: 16
များ: 17
တို့: 18
နဲ့: 19


In [102]:
!tail -n 20 ../vocab/word_vocab.yml

အေးတိအေးစက်: 6816
အဲဒီလောက်: 6817
အံ့သြ: 6818
ဣန္ဒြေ: 6819
ဣန္ဒြေပျက်: 6820
ဥပေက္ခာပြုထား: 6821
ဦးစားမပေး: 6822
ဦးဆောင်ပြီး: 6823
ဦးမော့: 6824
ဧကန်: 6825
ဧည့်: 6826
ေတွ: 6827
ေသာ: 6828
ကြပါ: 6829
ဖြစ်သည်: 6830
မကြိုးစား: 6831
မည်ဟု: 6832
မလိုပါ: 6833
မလိုဘူး: 6834
အသီးမထပါ: 6835


**Syllable normalization လုပ်လိုက်ရင်တော့ အောက်ပါလိုမျိုး ကိစ္စတွေ ပျောက်သွားလိမ့်မယ်။ သို့သော် syllable normalized training လုပ်ထားရင်၊ online testing လုပ်တဲ့အခါမှာလည်း syllable normalize လုပ်ပြီးမှ model ထဲကို input  လုပ်ရလိမ့်မယ်။**  

ေတွ: 6827  
ေသာ: 6828  

## Cleaning Train/Valid/Test Data



In [104]:
%cd /mnt/disk1/ye/exp/contra/data/

/mnt/disk1/ye/exp/contra/data


In [105]:
!mkdir syl

In [106]:
!cp /mnt/disk1/ye/exp/contra/myContradict/version1.0/syllable_segmented/* ./syl

In [107]:
%cd syl

/mnt/disk1/ye/exp/contra/data/syl


In [108]:
!ls

syl_test.src.txt  syl_train.src.txt  syl_valid.src.txt
syl_test.tgt.txt  syl_train.tgt.txt  syl_valid.tgt.txt


In [109]:
!cat syl_test.src.txt | python ../../scripts/clean-punctuation.py > test.src

In [110]:
!cat syl_test.tgt.txt | python ../../scripts/clean-punctuation.py > test.tgt

In [111]:
!cat syl_train.src.txt | python ../../scripts/clean-punctuation.py > train.src

In [112]:
!cat syl_train.tgt.txt | python ../../scripts/clean-punctuation.py > train.tgt

In [113]:
!cat syl_valid.src.txt | python ../../scripts/clean-punctuation.py > valid.src

In [114]:
!cat syl_valid.tgt.txt | python ../../scripts/clean-punctuation.py > valid.tgt

**Command ပေးတာမှာ အမှားအယွင်းမရှိအောင် ဂရုစိုက်ပါ။ ပြန်စစ်ပါ။**  

In [115]:
!wc *.src

   1214   17196  169851 test.src
   9702  133354 1315036 train.src
   1214   16458  162747 valid.src
  12130  167008 1647634 total


In [116]:
!wc *.tgt

   1214   17173  169507 test.tgt
   9702  133387 1315717 train.tgt
   1214   16452  162450 valid.tgt
  12130  167012 1647674 total


In [117]:
%pwd

'/mnt/disk1/ye/exp/contra/data/syl'


ကော်ပီကူးယူထားတဲ့ cleaning မလုပ်ရသေးတဲ့ ဖိုင်တွေကို ဖျက်ထားလိုက်ကြရအောင်။ 

In [118]:
!rm *.txt

In [119]:
!ls -lh *

-rw-rw-r-- 1 ye ye 166K Sep 26 09:10 test.src
-rw-rw-r-- 1 ye ye 166K Sep 26 09:10 test.tgt
-rw-rw-r-- 1 ye ye 1.3M Sep 26 09:11 train.src
-rw-rw-r-- 1 ye ye 1.3M Sep 26 09:11 train.tgt
-rw-rw-r-- 1 ye ye 159K Sep 26 09:12 valid.src
-rw-rw-r-- 1 ye ye 159K Sep 26 09:12 valid.tgt


In [162]:
%cd /mnt/disk1/ye/exp/contra/data/syl

/mnt/disk1/ye/exp/contra/data/syl


In [163]:
!tail test.src

သင်္ဘော နဲ့ ခ ရီး သွား ရ တာ တ ကယ် ကို ပျင်း ရိ စ ရာ ကောင်း ပါ တယ်
ထို သိုး မွေး မျိုး မှာ ရ လွယ် သ ဖြင့် အ လွန် ဈေး ချို သည်
အ ခြေ အ နေ ကြည့် ပြီး ကျွန် တော် တို့ က ခင် ဗျား တို့ ရဲ့ စ နစ် ကို လက် ခံ လိုက် တယ်
မ ကြိုက် ပါ သီ ချင်း ကို
ငွေ များ ရင် နိုင် တဲ့ ခတေ် ဖြစ် နေ တာ
မြို့ တွင်း
ကျွန် တော့် ပိုက် ဆံ ပြန် ယူ နိုင် မ လား
ဆ ရာ့ ရဲ့ အ ကြံ ပေး စ ကား တွေ က ကျွန် တော့် အ တွက် တန် ဖိုး ဖြတ် မ ရ အောင် ကောင်း လွန်း ခဲ့ တယ်
ကောင်း လိုက် တဲ့ အ တွေး အ ခေါ် ပါ
ကျွန် မ ကိုယ် ကျွန် မ အ လွန် ရှက် ခဲ့ မိ ပါ တယ်


In [164]:
!tail test.tgt

သင်္ဘော နဲ့ ခ ရီး သွား ရ တာ တ ကယ် ကို အံ့ ဩ စ ရာ ကောင်း ပါ တယ်
ထို သိုး မွေး မျိုး မှာ ရ တောင့် ရ ခဲ သ ဖြင့် အ လွန် အ ဖိုး ထိုက် တန် သည်
အ ခြေ အ နေ ကြည့် ပြီး ကျွန် တော် တို့ က ခင် ဗျား တို့ ရဲ့ စ နစ် ကို ငြင်း ပယ် လိုက် တယ်
ကြွေ သွား ပြီ သီ ချင်း ကို
ငွေ များ ရင် နိုင် တဲ့ ခတေ် မ ဟုတ် ဘူး
မြို့ ပြင်
ကျွန် တော့် ပိုက် ဆံ ပြန် ပေး နိုင် မ လား
ဆ ရာ့ ရဲ့ အ ကြံ ပေး စ ကား တွေ က ကျွန် တော့် အ တွက် တန် ဖိုး ဖြတ် မ ရ အောင် ဆိုး လွန်း ခဲ့ တယ်
ဆိုး ဆိုး ရွား ရွား အ တွေး အ ခေါ် ပါ
ကျွန် မ ကိုယ် ကျွန် မ အ လွန် မ ရှက် မိ ခဲ့ ပါ


In [165]:
!tail valid.src

မင်း ရဲ့ ကိစ္စ က ချွင်း ချက် မ ဟုတ် လို့ ပြော လို့ ရ တယ်
ထို ဂြိုဟ် သည် ဓာတ် ငွေ့ အ ပြည့် အ ဝ အ ခြေ ၌ သာ ရှိ ကောင်း ရှိ လိမ့် မည် ဟု သိပ္ပံ ပ ညာ ရှင် များ က ယူ ဆ ကြ လေ သည်
ကြို တင် ကာ ကွယ် ခြင်း သည် ကု သ ခြင်း ထက် ကောင်း ပါ သည်
ကိုယ့် ရဲ့ စာ နာ နား လည် မှု တွေ ကို မင်း လဈ် လျူ မ ပြု သင့် ဘူး ကွာ
အ မေ့ ဆီ က စာ ရ လား
ထုတ် ပိုး ပြီး ဖဲ ပြား နဲ့ မ ချည် ပေး လည်း ရ ပါ တယ်
မြင် နိုင် စွမ်း သော ပစ္စည်း ဖြစ် သည်
ငေါက် ငမ်း မ ပြော ဘဲ ချို ချို သာ သာ ပြော သင့် ပါ တယ်
တွဲ ကပ် သည်
ထို အ ချိန် မှ စ ၍ စာ ရေး ဆ ရာ နှင့် ဝတ္ထု အ တတ် ပ ညာ ကို စိတ် ဝင် စား ကာ စာ ရေး ဆ ရာ ဖြစ် ချင် ခဲ့ သည်


In [166]:
!tail valid.tgt

မင်း ရဲ့ ကိစ္စ က ချွင်း ချက် အ နေ နဲ့ ပြော လို့ မ ရ ဘူး
ထို ဂြိုဟ် သည် ဓာတ် ငွေ့ တစ် ဝက် တစ် ပျက် အ ခြေ ၌ သာ ရှိ ကောင်း ရှိ လိမ့် မည် ဟု သိပ္ပံ ပ ညာ ရှင် များ က ယူ ဆ ကြ လေ သည်
ကု သ ခြင်း သည် ကြို တင် ကာ ကွယ် ခြင်း လောက် မ ကောင်း ပါ
ကိုယ့် ရဲ့ စာ နာ နား လည် မှု တွေ ကို မင်း လဈ် လျူ ပြု လိုက် ပါ ကွာ
အ မေ့ ဆီ က စာ မ ရ ဘူး လား
ထုတ် ပိုး ပြီး ဖဲ ပြား နဲ့ ချည် ပေး ပါ
မ မြင် စွမ်း နိုင် သော ပစ္စည်း ဖြစ် သည်
ငေါက် ငမ်း ပြော ပြီး မာ မာ ထန် ထန် ပြော သင့် ပါ တယ်
ကွာ သည်
ထို အ ချိန် မှ စ ၍ စာ ရေး ဆ ရာ နှင့် ဝတ္ထု အ တတ် ပ ညာ ကို စိတ် မ ဝင် စား ပေ မဲ့ စာ ရေး ဆ ရာ ဖြစ် ချင် ခဲ့ သည်


In [167]:
!tail train.src

ကျန်း မာ ရေး နဲ့ မ ညီ ညွတ် တဲ့ အ စား အ စာ တွေ ကို သာ စား သုံး တယ်
ရ ထား က ရန် ကုန် ဘူ တာ ကြီး ကို ဆိုက် ရောက် မှာ ပါ
ပေး ထား ချက် များ ထက် ကျော် လွန် ၍ ဆင် ခြင် ခြင်း ကို ဝိ သေ သ ဆင် ခြင် ခြင်း ဟု ခေါ် လေ သည်
ငယ် ငယ် တုန်း က သင် ပုန်း နဲ့ စာ မ သင် ခဲ့ ရ ဖူး ဘူး
ကြား ရ တာ စိတ် မ ကောင်း ပါ ဘူး
ငါ့ ကို ဘာ ကြောင့် ညာ ရ သ လဲ
မ ကြည့် ချင် ဘူး တိုက် ရိုက် မ ထုတ် လွှင့် ပေး နဲ့ နော်
တူ ညီ မှု တွေ အ မြဲ တမ်း ရှိ နေ တယ်
သူ ဟာ အို လံ ပစ် ပြိုင် ပွဲ မှာ ရွှေ တံ ဆိပ် ဆု မ ရ ခဲ့ ဘူး
အ ချို့ ငါး များ သည် ရာ သီ အ လိုက် ပင် လယ် ကို စွန့် ကာ မြစ် ချောင်း များ ကို ဆန် တက် ပြီး ဥ များ ကို ဥ တတ် ကြ ၏


In [168]:
!tail train.tgt

ကျန်း မာ ရေး နဲ့ ညီ ညွတ် တဲ့ အ စား အ စာ တွေ ကို သာ စား သုံး တယ်
ရ ထား က ရန် ကုန် ဘူ တာ ကြီး က စ ထွက် မှာ ပါ
ပေး ထား ချက် များ ထက် ကျော် လွန် ခြင်း မ ပြု ဘဲ ဆင် ခြင် ခြင်း ကို ဝိ သေ သ ဆင် ခြင် ခြင်း ဟု ခေါ် သည်
ငယ် ငယ် တုန်း က သင် ပုန်း နဲ့ စာ သင် ခဲ့ ရ တယ်
ကြား ရ တာ ဝမ်း သာ ပါ ဘူး
ငါ့ ကို ဘာ ကြောင့် အ မှန် ပြော ရ သ လဲ
ကြည့် ချင် တယ် တိုက် ရိုက် ထုတ် လွှင့် ပေး ပါ နော်
ကွဲ ပြား ခြား နား မှု တွေ အ မြဲ တမ်း ရှိ နေ တယ်
သူ ဟာ အို လံ ပစ် ပြိုင် ပွဲ မှာ ရွှေ တံ ဆိပ် ဆု ရ ခဲ့ တယ်
အ ချို့ ငါး များ သည် ရာ သီ အ လိုက် ပင် လယ် ကို စွန့် ကာ မြစ် ချောင်း များ ကို စုန် တက် ပြီး ဥ များ ကို ဥ တတ် ကြ ၏


**ဒီတခါတော့ Word unit ဖြတ်ထားတဲ့ train/valid/test ဒေတာတွေကို cleaning လုပ်ကြရအောင်**

In [170]:
%pwd

'/mnt/disk1/ye/exp/contra/data/syl'

In [171]:
%cd ..

/mnt/disk1/ye/exp/contra/data


In [172]:
!mkdir word

mkdir: cannot create directory ‘word’: File exists


In [173]:
%cd word

/mnt/disk1/ye/exp/contra/data/word


In [125]:
!cp /mnt/disk1/ye/exp/contra/myContradict/version1.0/word_segmented/* .

In [126]:
!ls

word_test.src.txt  word_train.src.txt  word_valid.src.txt
word_test.tgt.txt  word_train.tgt.txt  word_valid.tgt.txt


In [145]:
!cat word_test.src.txt | python ../../scripts/clean-punctuation.py > test.src

In [146]:
!cat word_test.tgt.txt | python ../../scripts/clean-punctuation.py > test.tgt

In [147]:
!cat word_valid.src.txt | python ../../scripts/clean-punctuation.py > valid.src

In [148]:
!cat word_valid.tgt.txt | python ../../scripts/clean-punctuation.py > valid.tgt

In [150]:
!cat word_train.src.txt | python ../../scripts/clean-punctuation.py > train.src

In [149]:
!cat word_train.tgt.txt | python ../../scripts/clean-punctuation.py > train.tgt

In [151]:
!wc *.src

   1214   12185  164864 test.src
   9702   94503 1276614 train.src
   1214   11818  158146 valid.src
  12130  118506 1599624 total


In [152]:
!wc *.tgt

   1214   12189  164562 test.tgt
   9702   94501 1277248 train.tgt
   1214   11799  157833 valid.tgt
  12130  118489 1599643 total


In [155]:
!tail test.src

သင်္ဘော နဲ့ ခရီး သွား ရ တာ တကယ် ကို ပျင်းရိ စရာ ကောင်း ပါ တယ်
ထို သိုးမွေး မျိုး မှာ ရ လွယ် သဖြင့် အလွန် ဈေးချို သည်
အခြေအနေ ကြည့် ပြီး ကျွန်တော် တို့ က ခင်ဗျား တို့ ရဲ့ စနစ် ကို လက်ခံ လိုက် တယ်
မ ကြိုက် ပါ သီချင်း ကို
ငွေ များ ရင် နိုင် တဲ့ ခေတ် ဖြစ် နေ တာ
မြို့ တွင်း
ကျွန်တော့် ပိုက်ဆံ ပြန် ယူ နိုင် မလား
ဆရာ့ ရဲ့ အကြံပေး စကား တွေ က ကျွန်တော့် အတွက် တန်ဖိုးဖြတ် မ ရ အောင် ကောင်း လွန်း ခဲ့ တယ်
ကောင်း လိုက် တဲ့ အတွေးအခေါ် ပါ
ကျွန်မ ကိုယ် ကျွန်မ အလွန် ရှက် ခဲ့ မိ ပါ တယ်


In [156]:
!tail test.tgt

သင်္ဘော နဲ့ ခရီး သွား ရ တာ တကယ် ကို အံ့ဩ စရာ ကောင်း ပါ တယ်
ထို သိုးမွေး မျိုး မှာ ရ တောင့် ရ ခဲ သဖြင့် အလွန် အဖိုးထိုက်တန် သည်
အခြေအနေ ကြည့် ပြီး ကျွန်တော် တို့ က ခင်ဗျား တို့ ရဲ့ စနစ် ကို ငြင်းပယ် လိုက် တယ်
ကြွေသွား ပြီ သီချင်း ကို
ငွေ များ ရင် နိုင် တဲ့ ခေတ် မ ဟုတ် ဘူး
မြို့ ပြင်
ကျွန်တော့် ပိုက်ဆံ ပြန် ပေး နိုင် မလား
ဆရာ့ ရဲ့ အကြံပေး စကား တွေ က ကျွန်တော့် အတွက် တန်ဖိုးဖြတ် မ ရ အောင် ဆိုး လွန်း ခဲ့ တယ်
ဆိုးဆိုးရွားရွား အတွေးအခေါ် ပါ
ကျွန်မ ကိုယ် ကျွန်မ အလွန် မ ရှက် မိ ခဲ့ ပါ


In [157]:
!tail valid.src

မင်း ရဲ့ ကိစ္စ က ချွင်းချက် မ ဟုတ် လို့ ပြော လို့ ရ တယ်
ထို ဂြိုဟ် သည် ဓာတ်ငွေ့ အပြည့်အဝ အခြေ ၌ သာ ရှိ ကောင်း ရှိ လိမ့် မည် ဟု သိပ္ပံ ပညာရှင် များ က ယူဆ ကြ လေ သည်
ကြိုတင် ကာကွယ် ခြင်း သည် ကုသ ခြင်း ထက် ကောင်း ပါ သည်
ကိုယ့် ရဲ့ စာနာနားလည် မှု တွေ ကို မင်း လျစ်လျူ မ ပြု သင့် ဘူး ကွာ
အမေ့ ဆီ က စာ ရ လား
ထုတ်ပိုး ပြီး ဖဲပြား နဲ့ မ ချည် ပေး လည်း ရ ပါ တယ်
မြင် နိုင် စွမ်း သော ပစ္စည်း ဖြစ် သည်
ငေါက်ငမ်း မ ပြော ဘဲ ချိုချိုသာသာ ပြော သင့် ပါ တယ်
တွဲကပ် သည်
ထို အချိန် မှ စ ၍ စာရေးဆရာ နှင့် ဝတ္ထု အတတ်ပညာ ကို စိတ်ဝင်စား ကာ စာရေးဆရာ ဖြစ် ချင် ခဲ့ သည်


In [158]:
!tail valid.tgt

မင်း ရဲ့ ကိစ္စ က ချွင်းချက် အနေနဲ့ ပြော လို့ မ ရ ဘူး
ထို ဂြိုဟ် သည် ဓာတ်ငွေ့ တစ်ဝက်တစ်ပျက် အခြေ ၌ သာ ရှိ ကောင်း ရှိ လိမ့် မည် ဟု သိပ္ပံ ပညာရှင် များ က ယူဆ ကြ လေ သည်
ကုသ ခြင်း သည် ကြိုတင် ကာကွယ် ခြင်း လောက် မ ကောင်း ပါ
ကိုယ့် ရဲ့ စာနာနားလည် မှု တွေ ကို မင်း လျစ်လျူ ပြု လိုက် ပါ ကွာ
အမေ့ ဆီ က စာ မ ရ ဘူး လား
ထုတ်ပိုး ပြီး ဖဲပြား နဲ့ ချည် ပေး ပါ
မ မြင် စွမ်း နိုင် သော ပစ္စည်း ဖြစ် သည်
ငေါက်ငမ်း ပြော ပြီး မာမာထန်ထန် ပြော သင့် ပါ တယ်
ကွာ သည်
ထို အချိန် မှ စ ၍ စာရေးဆရာ နှင့် ဝတ္ထု အတတ်ပညာ ကို စိတ်မဝင်စား ပေမဲ့ စာရေးဆရာ ဖြစ် ချင် ခဲ့ သည်


In [160]:
!tail train.src

ကျန်းမာ ရေး နဲ့ မညီညွတ် တဲ့ အစားအစာ တွေ ကို သာ စား သုံး တယ်
ရထား က ရန်ကုန် ဘူတာ ကြီး ကို ဆိုက်ရောက် မှာ ပါ
ပေးထားချက် များ ထက် ကျော်လွန် ၍ ဆင်ခြင် ခြင်း ကို ဝိသေသ ဆင်ခြင် ခြင်း ဟု ခေါ် လေ သည်
ငယ်ငယ် တုန်း က သင်ပုန်း နဲ့ စာ မ သင် ခဲ့ ရ ဖူး ဘူး
ကြား ရ တာ စိတ်မကောင်း ပါ ဘူး
ငါ့ ကို ဘာ ကြောင့် ညာ ရ သလဲ
မ ကြည့် ချင် ဘူး တိုက်ရိုက် မ ထုတ်လွှင့်ပေး နဲ့ နော်
တူညီ မှု တွေ အမြဲတမ်း ရှိ နေ တယ်
သူ ဟာ အိုလံပစ် ပြိုင်ပွဲ မှာ ရွှေ တံဆိပ် ဆု မ ရ ခဲ့ ဘူး
အချို့ ငါး များ သည် ရာသီ အလိုက် ပင်လယ် ကို စွန့် ကာ မြစ် ချောင်း များ ကို ဆန်တက် ပြီး ဥ များ ကို ဥ တတ် ကြ ၏


In [161]:
!tail train.tgt

ကျန်းမာ ရေး နဲ့ ညီညွတ် တဲ့ အစားအစာ တွေ ကို သာ စား သုံး တယ်
ရထား က ရန်ကုန် ဘူတာ ကြီး က စ ထွက် မှာ ပါ
ပေးထားချက် များ ထက် ကျော်လွန် ခြင်း မ ပြု ဘဲ ဆင်ခြင် ခြင်းကို ဝိသေသ ဆင်ခြင် ခြင်း ဟု ခေါ် သည်
ငယ်ငယ် တုန်း က သင်ပုန်း နဲ့ စာသင် ခဲ့ ရ တယ်
ကြား ရ တာ ဝမ်းသာ ပါ ဘူး
ငါ့ ကို ဘာ ကြောင့် အမှန်ပြော ရ သလဲ
ကြည့် ချင် တယ် တိုက်ရိုက် ထုတ်လွှင့်ပေး ပါ နော်
ကွဲပြားခြားနား မှု တွေ အမြဲတမ်း ရှိ နေ တယ်
သူ ဟာ အိုလံပစ် ပြိုင်ပွဲ မှာ ရွှေ တံဆိပ် ဆု ရ ခဲ့ တယ်
အချို့ ငါး များ သည် ရာသီ အလိုက် ပင်လယ် ကို စွန့် ကာ မြစ် ချောင်း များ ကို စုန်တက် ပြီး ဥ များ ကို ဥ တတ် ကြ ၏


In [176]:
%rm *.txt

In [177]:
!tree /mnt/disk1/ye/exp/contra/data

/mnt/disk1/ye/exp/contra/data
├── preprocessing
│   ├── syl_all.txt
│   ├── syl_all.txt.clean
│   ├── word_all.txt
│   └── word_all.txt.clean
├── syl
│   ├── test.src
│   ├── test.tgt
│   ├── train.src
│   ├── train.tgt
│   ├── valid.src
│   └── valid.tgt
├── vocab
│   ├── syl_vocab.yml
│   └── word_vocab.yml
└── word
    ├── test.src
    ├── test.tgt
    ├── train.src
    ├── train.tgt
    ├── valid.src
    └── valid.tgt

5 directories, 18 files


## Preparing Bash Scripts for Sequence to Sequence Model



In [178]:
%cd /mnt/disk1/ye/exp/contra

/mnt/disk1/ye/exp/contra


In [185]:
!cat ./s2s.syl.sh

#!/bin/bash

## Written by Ye Kyaw Thu, Language Understanding Lab, AI Mario, Myanmar
## NMT Tutorial for AI Engineering Class (Batch-2)
## Date: 26 Sept 2026

## Reference: https://marian-nmt.github.io/examples/mtm2017/complex/

model_folder="model.s2s.syl";
mkdir -p ${model_folder};
data_path="/mnt/disk1/ye/exp/contra/data/syl/";
vocab_path="/mnt/disk1/ye/exp/contra/data/vocab/"
src="src"; tgt="tgt";


marian \
  --type s2s \
  --train-sets ${data_path}/train.${src} ${data_path}/train.${tgt} \
  --max-length 200 \
  --valid-sets ${data_path}/valid.${src} ${data_path}/valid.${tgt} \
  --vocabs  ${vocab_path}/syl_vocab.yml  ${vocab_path}/syl_vocab.yml \
  --model ${model_folder}/model.npz \
  --workspace 500 \
  --enc-depth 2 --enc-type alternating --enc-cell lstm --enc-cell-depth 2 \
  --dec-depth 2 --dec-cell lstm --dec-cell-base-depth 2 --dec-cell-high-depth 2 \
  --tied-embeddings --layer-normalization --skip \
  --mini-batch-fit \
  --valid-mini-batch 32 \
  --valid-metrics cross-

## Check GPU Status

In [186]:
!nvidia-smi

Sat Sep 26 10:24:13 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 595.84                 Driver Version: 595.84         CUDA Version: 13.2     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3090 Ti     Off |   00000000:01:00.0 Off |                  Off |
|  0%   50C    P8             27W /  480W |    3821MiB /  24564MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

## Training for Syllable Unit Data

In [187]:
!time ./s2s.syl.sh

[2026-09-26 10:24:16] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 10:24:16] [marian] Running on lst-hpc3090 as process 1505305 with command line:
[2026-09-26 10:24:16] [marian] marian -c model.s2s.syl/src-tgt.config.yml
[2026-09-26 10:24:16] [config] after: 0e
[2026-09-26 10:24:16] [config] after-batches: 0
[2026-09-26 10:24:16] [config] after-epochs: 0
[2026-09-26 10:24:16] [config] all-caps-every: 0
[2026-09-26 10:24:16] [config] allow-unk: false
[2026-09-26 10:24:16] [config] authors: false
[2026-09-26 10:24:16] [config] beam-size: 12
[2026-09-26 10:24:16] [config] bert-class-symbol: "[CLS]"
[2026-09-26 10:24:16] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 10:24:16] [config] bert-masking-fraction: 0.15
[2026-09-26 10:24:16] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 10:24:16] [config] bert-train-type-embeddings: true
[2026-09-26 10:24:16] [config] bert-type-vocab-size: 2
[2026-09-26 10:24:16] [config] build-info: ""
[2026-09-26 10:24:16] [config] ch

In [189]:
%cd /mnt/disk1/ye/exp/contra/model.s2s.syl/

/mnt/disk1/ye/exp/contra/model.s2s.syl


In [190]:
!ls

model.iter10000.npz  model.iter45000.npz    model.npz.optimizer.npz
model.iter15000.npz  model.iter50000.npz    model.npz.progress.yml
model.iter20000.npz  model.iter5000.npz     model.npz.yml
model.iter25000.npz  model.iter55000.npz    s2s.src-tgt.log
model.iter30000.npz  model.iter60000.npz    src-tgt.config.yml
model.iter35000.npz  model.npz		    train.log
model.iter40000.npz  model.npz.decoder.yml  valid.log


## Testing with S2S Model

In [196]:
%cd /mnt/disk1/ye/exp/contra/model.s2s.syl

/mnt/disk1/ye/exp/contra/model.s2s.syl


In [197]:
!cat test-eval.sh

#!/bin/bash

for i in {5000..60000..5000}
do
    marian-decoder -m ./model.iter$i.npz -v /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml --devices 0 --output hyp.iter$i.tgt < /mnt/disk1/ye/exp/contra/data/syl/test.src;
    echo "Evaluation with hyp.iter$i.tgt, Transformer model:" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/contra/data/syl/test.tgt < ./hyp.iter$i.tgt >> eval-result.txt;

done


In [ ]:
!time ./test-eval.sh

[2026-09-26 18:04:28] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 18:04:28] [marian] Running on lst-hpc3090 as process 1614905 with command line:
[2026-09-26 18:04:28] [marian] marian-decoder -m ./model.iter5000.npz -v /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml --devices 0 --output hyp.iter5000.tgt
[2026-09-26 18:04:28] [config] alignment: ""
[2026-09-26 18:04:28] [config] allow-special: false
[2026-09-26 18:04:28] [config] allow-unk: false
[2026-09-26 18:04:28] [config] authors: false
[2026-09-26 18:04:28] [config] beam-size: 12
[2026-09-26 18:04:28] [config] bert-class-symbol: "[CLS]"
[2026-09-26 18:04:28] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 18:04:28] [config] bert-masking-fraction: 0.15
[2026-09-26 18:04:28] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 18:04:28] [config] bert-train-type-embeddings: true
[2026-09-26 18:04:28] [config] bert-type-vocab-size: 2
[2026-09-26 18:04:28] [config]

In [204]:
!cat eval-result.txt

Evaluation with hyp.iter5000.tgt, S2S model:
BLEU = 50.76, 77.7/62.9/51.8/43.1 (BP=0.883, ratio=0.889, hyp_len=15270, ref_len=17173)
Evaluation with hyp.iter10000.tgt, S2S model:
BLEU = 68.93, 84.3/74.0/65.9/59.2 (BP=0.981, ratio=0.981, hyp_len=16851, ref_len=17173)
Evaluation with hyp.iter15000.tgt, S2S model:
BLEU = 70.31, 84.9/74.9/67.1/60.7 (BP=0.986, ratio=0.986, hyp_len=16926, ref_len=17173)
Evaluation with hyp.iter20000.tgt, S2S model:
BLEU = 69.57, 84.2/74.0/66.3/60.0 (BP=0.986, ratio=0.986, hyp_len=16937, ref_len=17173)
Evaluation with hyp.iter25000.tgt, S2S model:
BLEU = 69.06, 83.9/73.6/66.0/59.8 (BP=0.983, ratio=0.983, hyp_len=16887, ref_len=17173)
Evaluation with hyp.iter30000.tgt, S2S model:
BLEU = 68.41, 83.3/73.0/65.3/59.2 (BP=0.982, ratio=0.982, hyp_len=16871, ref_len=17173)
Evaluation with hyp.iter35000.tgt, S2S model:
BLEU = 68.19, 82.9/72.7/65.1/59.1 (BP=0.983, ratio=0.983, hyp_len=16878, ref_len=17173)
Evaluation with hyp.iter40000.tgt, S2S model:
BLEU = 67.96, 82.

## Prepare Bash Shell Script for Transformer Model Building

In [192]:
%cd /mnt/disk1/ye/exp/contra/

/mnt/disk1/ye/exp/contra


```bash
#!/bin/bash

## Written by Ye Kyaw Thu, Language Understanding Lab, AI Mario, Myanmar
## Experiments for AI Engineering Fundamental Class (Batch-2)
## Date: 26 Sept 2026

## Old Notes
#     --mini-batch-fit -w 10000 --maxi-batch 1000 \
#    --mini-batch-fit -w 1000 --maxi-batch 100 \
#     --tied-embeddings-all \
#     --tied-embeddings \
#     --valid-metrics cross-entropy perplexity translation bleu \
#     --transformer-dropout 0.1 --label-smoothing 0.1 \
#     --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report \
#     --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \

model_folder="model.tf.syl";
mkdir -p ${model_folder};
data_path="/mnt/disk1/ye/exp/contra/data/syl/";
vocab_path="/mnt/disk1/ye/exp/contra/data/vocab/"
src="src"; tgt="tgt";

marian \
    --model ${model_folder}/model.npz --type transformer \
    --train-sets ${data_path}/train.${src} ${data_path}/train.${tgt} \
    --max-length 200 \
    --vocabs  ${vocab_path}/word_vocab.yml  ${vocab_path}/word_vocab.yml \
    --mini-batch-fit -w 1000 --maxi-batch 100 \
    --early-stopping 10 \
    --valid-freq 5000 --save-freq 5000 --disp-freq 500 \
    --valid-metrics cross-entropy perplexity bleu \
    --valid-sets ${data_path}/dev.${src} ${data_path}/dev.${tgt} \
    --valid-translation-output ${model_folder}/valid.${src}-${tgt}.output \
    --quiet-translation \
    --valid-mini-batch 128 \
    --beam-size 6 --normalize 0.6 \
    --log ${model_folder}/train.log \
    --valid-log ${model_folder}/valid.log \
    --enc-depth 2 --dec-depth 2 \
    --transformer-heads 8 \
    --transformer-postprocess-emb d \
    --transformer-postprocess dan \
    --transformer-dropout 0.3 --label-smoothing 0.1 \
    --learn-rate 0.0003 --lr-warmup 0 --lr-decay-inv-sqrt 16000 --lr-report \
    --clip-norm 5 \
    --tied-embeddings \
    --devices 0 --sync-sgd --seed 1111 \
    --exponential-smoothing \
    --dump-config > ${model_folder}/${src}-${tgt}.config.yml

time marian -c ${model_folder}/${src}-${tgt}.config.yml 2>&1 | tee ${model_folder}/tf.${src}-${tgt}.log
```

## Training Transformer Model

In [193]:
!./tf.syl.sh

[2026-09-26 17:22:54] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 17:22:54] [marian] Running on lst-hpc3090 as process 1605810 with command line:
[2026-09-26 17:22:54] [marian] marian -c model.tf.syl/src-tgt.config.yml
[2026-09-26 17:22:54] [config] after: 0e
[2026-09-26 17:22:54] [config] after-batches: 0
[2026-09-26 17:22:54] [config] after-epochs: 0
[2026-09-26 17:22:54] [config] all-caps-every: 0
[2026-09-26 17:22:54] [config] allow-unk: false
[2026-09-26 17:22:54] [config] authors: false
[2026-09-26 17:22:54] [config] beam-size: 6
[2026-09-26 17:22:54] [config] bert-class-symbol: "[CLS]"
[2026-09-26 17:22:54] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 17:22:54] [config] bert-masking-fraction: 0.15
[2026-09-26 17:22:54] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 17:22:54] [config] bert-train-type-embeddings: true
[2026-09-26 17:22:54] [config] bert-type-vocab-size: 2
[2026-09-26 17:22:54] [config] build-info: ""
[2026-09-26 17:22:54] [config] chec

## Check GPU Status

In [194]:
!nvidia-smi

Sat Sep 26 17:23:55 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 595.84                 Driver Version: 595.84         CUDA Version: 13.2     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3090 Ti     Off |   00000000:01:00.0 Off |                  Off |
|  0%   47C    P8             27W /  480W |    3821MiB /  24564MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

## Error

```
 Error: File '/mnt/disk1/ye/exp/contra/data/syl//dev.src' does not exist
```

In [195]:
!./tf.syl.sh

[2026-09-26 17:26:34] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 17:26:34] [marian] Running on lst-hpc3090 as process 1606569 with command line:
[2026-09-26 17:26:34] [marian] marian -c model.tf.syl/src-tgt.config.yml
[2026-09-26 17:26:34] [config] after: 0e
[2026-09-26 17:26:34] [config] after-batches: 0
[2026-09-26 17:26:34] [config] after-epochs: 0
[2026-09-26 17:26:34] [config] all-caps-every: 0
[2026-09-26 17:26:34] [config] allow-unk: false
[2026-09-26 17:26:34] [config] authors: false
[2026-09-26 17:26:34] [config] beam-size: 6
[2026-09-26 17:26:34] [config] bert-class-symbol: "[CLS]"
[2026-09-26 17:26:34] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 17:26:34] [config] bert-masking-fraction: 0.15
[2026-09-26 17:26:34] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 17:26:34] [config] bert-train-type-embeddings: true
[2026-09-26 17:26:34] [config] bert-type-vocab-size: 2
[2026-09-26 17:26:34] [config] build-info: ""
[2026-09-26 17:26:34] [config] chec

## Testing Transformer Model 

In [205]:
%cd /mnt/disk1/ye/exp/contra/model.tf.syl

/mnt/disk1/ye/exp/contra/model.tf.syl


In [207]:
!cat test-eval.sh

#!/bin/bash

for i in {5000..55000..5000}
do
    marian-decoder -m ./model.iter$i.npz -v /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml --devices 0 --output hyp.iter$i.tgt < /mnt/disk1/ye/exp/contra/data/syl/test.src;
    echo "Evaluation with hyp.iter$i.tgt, Transformer model:" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/contra/data/syl/test.tgt < ./hyp.iter$i.tgt >> eval-result.txt;

done


In [206]:
!./test-eval.sh

[2026-09-26 18:28:58] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 18:28:58] [marian] Running on lst-hpc3090 as process 1620457 with command line:
[2026-09-26 18:28:58] [marian] marian-decoder -m ./model.iter5000.npz -v /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml --devices 0 --output hyp.iter5000.tgt
[2026-09-26 18:28:58] [config] alignment: ""
[2026-09-26 18:28:58] [config] allow-special: false
[2026-09-26 18:28:58] [config] allow-unk: false
[2026-09-26 18:28:58] [config] authors: false
[2026-09-26 18:28:58] [config] beam-size: 12
[2026-09-26 18:28:58] [config] bert-class-symbol: "[CLS]"
[2026-09-26 18:28:58] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 18:28:58] [config] bert-masking-fraction: 0.15
[2026-09-26 18:28:58] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 18:28:58] [config] bert-train-type-embeddings: true
[2026-09-26 18:28:58] [config] bert-type-vocab-size: 2
[2026-09-26 18:28:58] [config]

## Debugging

```
    --vocabs  ${vocab_path}/word_vocab.yml  ${vocab_path}/word_vocab.yml \
```

I updated as follows:

```
    --vocabs  ${vocab_path}/syl_vocab.yml  ${vocab_path}/syl_vocab.yml \
```

## ReTraining Transformer Model

In [210]:
%cd /mnt/disk1/ye/exp/contra

/mnt/disk1/ye/exp/contra


In [211]:
!./tf.syl.sh

[2026-09-26 18:42:40] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 18:42:40] [marian] Running on lst-hpc3090 as process 1623710 with command line:
[2026-09-26 18:42:40] [marian] marian -c model.tf.syl/src-tgt.config.yml
[2026-09-26 18:42:40] [config] after: 0e
[2026-09-26 18:42:40] [config] after-batches: 0
[2026-09-26 18:42:40] [config] after-epochs: 0
[2026-09-26 18:42:40] [config] all-caps-every: 0
[2026-09-26 18:42:40] [config] allow-unk: false
[2026-09-26 18:42:40] [config] authors: false
[2026-09-26 18:42:40] [config] beam-size: 6
[2026-09-26 18:42:40] [config] bert-class-symbol: "[CLS]"
[2026-09-26 18:42:40] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 18:42:40] [config] bert-masking-fraction: 0.15
[2026-09-26 18:42:40] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 18:42:40] [config] bert-train-type-embeddings: true
[2026-09-26 18:42:40] [config] bert-type-vocab-size: 2
[2026-09-26 18:42:40] [config] build-info: ""
[2026-09-26 18:42:40] [config] chec

## Let's See Models

In [213]:
!tree ./model.tf.syl/

./model.tf.syl/
├── model.iter10000.npz
├── model.iter15000.npz
├── model.iter20000.npz
├── model.iter25000.npz
├── model.iter30000.npz
├── model.iter35000.npz
├── model.iter40000.npz
├── model.iter45000.npz
├── model.iter50000.npz
├── model.iter5000.npz
├── model.iter55000.npz
├── model.npz
├── model.npz.decoder.yml
├── model.npz.optimizer.npz
├── model.npz.progress.yml
├── model.npz.yml
├── src-tgt.config.yml
├── tf.src-tgt.log
├── train.log
├── valid.log
└── valid.src-tgt.output

1 directory, 21 files


In [214]:
!ls -lh ./model.tf.syl/*.npz

-rw-rw-r-- 1 ye ye  63M Sep 26 18:45 ./model.tf.syl/model.iter10000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:46 ./model.tf.syl/model.iter15000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:48 ./model.tf.syl/model.iter20000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:49 ./model.tf.syl/model.iter25000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:51 ./model.tf.syl/model.iter30000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:52 ./model.tf.syl/model.iter35000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:54 ./model.tf.syl/model.iter40000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:55 ./model.tf.syl/model.iter45000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:57 ./model.tf.syl/model.iter50000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:44 ./model.tf.syl/model.iter5000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:58 ./model.tf.syl/model.iter55000.npz
-rw-rw-r-- 1 ye ye  63M Sep 26 18:58 ./model.tf.syl/model.npz
-rw-rw-r-- 1 ye ye 251M Sep 26 18:58 ./model.tf.syl/model.npz.optimizer.npz


In [215]:
!ls -lh ./model.s2s.syl/*.npz

-rw-rw-r-- 1 ye ye 343M Sep 26 10:35 ./model.s2s.syl/model.iter10000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 10:41 ./model.s2s.syl/model.iter15000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 10:46 ./model.s2s.syl/model.iter20000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 10:52 ./model.s2s.syl/model.iter25000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 10:58 ./model.s2s.syl/model.iter30000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:03 ./model.s2s.syl/model.iter35000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:09 ./model.s2s.syl/model.iter40000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:15 ./model.s2s.syl/model.iter45000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:20 ./model.s2s.syl/model.iter50000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 10:29 ./model.s2s.syl/model.iter5000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:26 ./model.s2s.syl/model.iter55000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:32 ./model.s2s.syl/model.iter60000.npz
-rw-rw-r-- 1 ye ye 343M Sep 26 11:32 ./model.s2s.syl/model.npz
-rw-rw-r-- 1 ye ye 1.4G Sep 26 11:32 ./model.s2s.syl/model.npz

## Testing Transformer Model

In [216]:
%cd /mnt/disk1/ye/exp/contra/model.tf.syl

/mnt/disk1/ye/exp/contra/model.tf.syl


In [217]:
!cat test-eval.sh

#!/bin/bash

for i in {5000..55000..5000}
do
    marian-decoder -m ./model.iter$i.npz -v /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml --devices 0 --output hyp.iter$i.tgt < /mnt/disk1/ye/exp/contra/data/syl/test.src;
    echo "Evaluation with hyp.iter$i.tgt, Transformer model:" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/contra/data/syl/test.tgt < ./hyp.iter$i.tgt >> eval-result.txt;

done


In [ ]:
!time ./test-eval.sh

[2026-09-26 19:02:34] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 19:02:34] [marian] Running on lst-hpc3090 as process 1627858 with command line:
[2026-09-26 19:02:34] [marian] marian-decoder -m ./model.iter5000.npz -v /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml /mnt/disk1/ye/exp/contra/data/vocab/syl_vocab.yml --devices 0 --output hyp.iter5000.tgt
[2026-09-26 19:02:34] [config] alignment: ""
[2026-09-26 19:02:34] [config] allow-special: false
[2026-09-26 19:02:34] [config] allow-unk: false
[2026-09-26 19:02:34] [config] authors: false
[2026-09-26 19:02:34] [config] beam-size: 12
[2026-09-26 19:02:34] [config] bert-class-symbol: "[CLS]"
[2026-09-26 19:02:34] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 19:02:34] [config] bert-masking-fraction: 0.15
[2026-09-26 19:02:34] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 19:02:34] [config] bert-train-type-embeddings: true
[2026-09-26 19:02:34] [config] bert-type-vocab-size: 2
[2026-09-26 19:02:34] [config]

## Check BLEU Scores

In [220]:
!cat eval-result.txt

Evaluation with hyp.iter5000.tgt, Transformer model:
BLEU = 72.57, 87.2/78.0/70.6/64.4 (BP=0.973, ratio=0.973, hyp_len=16715, ref_len=17173)
Evaluation with hyp.iter10000.tgt, Transformer model:
BLEU = 74.10, 87.4/78.5/71.4/65.5 (BP=0.984, ratio=0.985, hyp_len=16907, ref_len=17173)
Evaluation with hyp.iter15000.tgt, Transformer model:
BLEU = 74.57, 87.5/78.7/71.7/65.9 (BP=0.987, ratio=0.987, hyp_len=16955, ref_len=17173)
Evaluation with hyp.iter20000.tgt, Transformer model:
BLEU = 74.91, 87.7/79.0/72.2/66.5 (BP=0.986, ratio=0.986, hyp_len=16940, ref_len=17173)
Evaluation with hyp.iter25000.tgt, Transformer model:
BLEU = 74.99, 87.6/79.0/72.1/66.4 (BP=0.988, ratio=0.988, hyp_len=16973, ref_len=17173)
Evaluation with hyp.iter30000.tgt, Transformer model:
BLEU = 75.00, 87.8/79.1/72.3/66.7 (BP=0.986, ratio=0.986, hyp_len=16934, ref_len=17173)
Evaluation with hyp.iter35000.tgt, Transformer model:
BLEU = 75.21, 88.0/79.4/72.6/67.0 (BP=0.985, ratio=0.985, hyp_len=16914, ref_len=17173)
Evaluat

## Compare S2S vs Transformer Model Result


| Iteration | S2S BLEU | Transformer BLEU | Difference / Transformer − S2S |
|---:|---:|---:|---:|
| 5000 | 50.76 | 72.57 | +21.81 |
| 10000 | 68.93 | 74.10 | +5.17 |
| 15000 | 70.31 | 74.57 | +4.26 |
| 20000 | 69.57 | 74.91 | +5.34 |
| 25000 | 69.06 | 74.99 | +5.93 |
| 30000 | 68.41 | 75.00 | +6.59 |
| 35000 | 68.19 | 75.21 | +7.02 |
| 40000 | 67.96 | 75.26 | +7.30 |
| 45000 | 67.91 | 75.22 | +7.31 |
| 50000 | 67.62 | 75.39 | +7.77 |
| 55000 | 67.24 | 75.54 | +8.30 |
| 60000 | 66.93 | — | — |
| **Best** | **70.31 @ 15000** | **75.54 @ 55000** | **+5.23 best-vs-best** |

## Best checkpoint comparison

| Model | Best iteration | BLEU | 1-gram | 2-gram | 3-gram | 4-gram | BP | ratio | hyp_len | ref_len |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| S2S | 15000 | **70.31** | 84.9 | 74.9 | 67.1 | 60.7 | 0.986 | 0.986 | 16926 | 17173 |
| Transformer | 55000 | **75.54** | 88.0 | 79.6 | 72.9 | 67.3 | 0.987 | 0.987 | 16946 | 17173 |
